asynciterator.ts

```diff
@@ -1981,42 +1978,172 @@ class HistoryReader<T> {
   }
 }
 
+/**
+ * An IterableWrapper iterates over native Iterable or Iterator instances.
+ */
+export class IterableWrapper<T> extends AsyncIterator<T> {
+  private _source?: Iterator<T>;
+
+  constructor(source: Iterator<T> | Iterable<T>) {
+    super();
+    this._source = isIterable(source) ? source[Symbol.iterator]() : source;
+    this.readable = true;
+  }
+
+  read(): T | null {
+    if (this._source) {
+      const item = this._source.next();
+      if (!item.done)
+        return item.value;
+      this.close();
+    }
+    return null;
+  }
+
+  protected _end(destroy: boolean = false) {
+    super._end(destroy);
+    delete this._source;
+  }
+}
+
+/**
+ * The WrappingIterator class can be used to wrap all supported kinds of data
+ * sources, both promisified and not, into AsyncIterator instances.
+ * This class is not meant to be used directly but, rather, through the
+ * `wrap()` function as it provides the mechanism that allows the latter
+ * to synchronously return an instance of AsyncIterator even for promisified
+ * sources.
+ */
+export class WrappingIterator<T> extends AsyncIterator<T> {
+  protected _source?: InternalSource<T>;
+
+  constructor(source?: MaybePromise<IterableSource<T>> | null) {
+    super();
+    if (source) {
+      if (!isPromise(source))
+        this._wrapSource(source as any);
+      else
+        source.then(s => this._wrapSource(s as any)).catch(e => this.emit('error', e));
+    }
+  }
+
+  protected _wrapSource(source: InternalSource<T>) {
+    try {
+      this._source = isAsyncReadable(source) ? source : fromSource(source);
+    }
+    catch (error) {
+      scheduleTask(() => this.emit('error', error));
+      return;
+    }
+    this._source._destination = this;
+    this._source.on('end', destinationClose);
+    this._source.on('error', destinationEmitError);
+    this._source.on('readable', () => this.emit('readable'));
```

Comment:
@jacoscaz This line shouldn't be here, but there are errors when I remove it. Could you please have a look? (The relevant commit is 555d88e043797df8b9f4262779b13d7c4ed5734c)

Reply:
@RubenVerborgh Looking into it. As a more general consideration, I don't think using generic listeners here is a good idea. If `isAsyncReadable(source)` returns true, we could still be dealing with an externally-provided, async-ish object such as a `stream.Readable` and IMHO in that case we shouldn't add the `_destination` property to it. This is one of reasons why I had opted to use dedicated event listeners as opposed to the generic ones.

Reply:
Seems to be something related to the debouncing of the `readable` event that happens at https://github.com/RubenVerborgh/AsyncIterator/blob/555d88e043797df8b9f4262779b13d7c4ed5734c/asynciterator.ts#L253-L262 . Commenting the `if` out makes the two tests that are failing pass (whilst breaking ~50 others).

Reply:
This works:

```js
    it('should emit all items', async () => {
      return new Promise((resolve) => {
        const items = [];
        iterator.once('end', () => {
          expect(items).to.deep.equal([0, 1, 2, 3]);
          resolve();
        });
        // iterator.on('readable', () => {
          let item;
          while ((item = iterator.read()) !== null) {
            items.push(item);
            console.log('ITEM', item);
          }
        // });
    });
```

Whereas it doesn't work if I uncomment the `readable` listener registration.

Reply:
Perfect thanks; then it's solved indeed, the bug was in the test. The readable event is and should only be fired when a non-readable iterator becomes readable.

Regarding generic listeners: it's a trick I picked up from the Node.js source a long time ago; the idea was to minimize object creation in performance-critical code. However, this advice might or might not be outdated. In any case, the rest of AsyncIterator is doing it, so I will stick to it for consistency. The _destination approach could be changed in the next major version with a WeakMap, or indeed with new listener objects every time.

Reply:
> Perfect thanks; then it's solved indeed, the bug was in the test. The readable event is and should only be fired when a non-readable iterator becomes readable.

Was it? the `toArray()` method should always work, regardless of whether it's invoked when the iterator is readable or not. I could very well be missing something but this looks still unaddressed to me.

Reply:
> so I will stick to it for consistency

Consistency is one of the many reasons why I like working on `asynciterator` so much. I made your [Programming is an art](https://ruben.verborgh.org/blog/2013/02/21/programming-is-an-art/) post mandatory reading in more than one occasion. Still, I don't think we should modify externally provided objects. Chances of collision in property names are unlikely but if one such collision were to happen it would lead to an extremely hard-to-debug behavior.

Reply:
@RubenVerborgh found the culprit and fixed in https://github.com/RubenVerborgh/AsyncIterator/pull/63/commits/159ed2b09e599d03d574aec2c98b40ae416d168b . Basically, I was ignoring the fact that a `null` return by `WrappingIterator#read()` should trigger a switch to `readable === false`, so that subsequent `readable` events from the source actually result in a `readable` event from the wrapper.

Reply:
> Collisions of property names are unlikely but not impossible and if one such collision were to happen it would lead to an extremely hard-to-debug behavior.

@RubenVerborgh thinking about it, a non-breaking workaround for this would be to rename the `_destination` property to a `Symbol()` (see the [relevant MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)). That way the chances of collision become virtually non-existent.

Reply:
Gave it a go in https://github.com/RubenVerborgh/AsyncIterator/pull/78 .

Reply:
Perfect, thanks for both of these!